stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -1522,21 +2144,19 @@ class LowerType
     if (auto *ReferenceTI = dyn_cast<ReferenceTypeInfo>(TI))
       return TC.getReferenceTypeInfo(Kind, ReferenceTI->getReferenceCounting());
 
+    if (auto *EnumTI = dyn_cast<EnumTypeInfo>(TI)) {
+      if (EnumTI->isOptional() && Kind == ReferenceKind::Weak) {
```

Comment:
Note the previous version of this code just used `SubKind == SinglePayloadEnum` as a test for something being an Optional.  I've factored that out in order to make it easier to refine this decision.